### PR TITLE
tests/xtimer_mutex_lock_timeout: New test

### PR DESCRIPTION
**Pull request by JulianHolzwarth**
Tracking: pr #11660

### Contribution description

Implements a new function for `xtimer_mutex_lock_timeout` to make sure the function works with threads. The new function tests `xtimer_mutex_lock_timeout` by locking the mutex from another thread, calling `xtimer_mutex_lock_timeout` and unlocking the mutex from the other thread before the timeout ends. 

(The test also shows that the created thread exits/ends)

The test also prints a empty line after each test function to make it easier to read and removes a wrong include.

### Testing procedure

`BOARD=native make -C tests/xtimer_mutex_lock_timeout/ flash test`
it outputs:
```
...
> mutex_timeout_long_unlocked
mutex_timeout_long_unlocked
starting test: xtimer mutex lock timeout
OK

> mutex_timeout_long_locked
mutex_timeout_long_locked
starting test: xtimer mutex lock timeout
OK

> mutex_timeout_long_locked_low
mutex_timeout_long_locked_low
starting test: xtimer mutex lock timeout with thread
threads = 2
THREAD low prio: start
MAIN THREAD: calling xtimer_mutex_lock_timeout
OK
threads = 3
MAIN THREAD: waiting for created thread to end
THREAD low prio: exiting low
threads = 2

> 
```

### Issues/PRs references

tracking #11660

### CODE REVIEW

File: tests/xtimer_mutex_lock_timeout/main.c [hunk 1]
```diff
@@ -42,9 +49,38 @@ static const shell_command_t shell_commands[] = {
...
+      "locked mutex from lower prio thread function with long timeout",
+      cmd_test_xtimer_mutex_lock_timeout_low_prio_thread, },
     { NULL, NULL, NULL }
 };
 
+/**
+ * @brief   thread function and stack for
+ *          cmd_test_xtimer_mutex_lock_timeout_low_prio_thread
+ */
+
```
Comment (MichelRottleuthner) [inline tests/xtimer_mutex_lock_timeout/main.c]: please remove empty line between doc and variable declaration

File: tests/xtimer_mutex_lock_timeout/main.c [hunk 2]
```diff
@@ -42,9 +49,38 @@ static const shell_command_t shell_commands[] = {
...
+void *thread_low_prio_test(void *arg)
+{
+    mutex_t *test_mutex = (mutex_t *)arg;
+    msg_t msg;
+
+    puts("THREAD low prio: start");
+
+    mutex_lock(test_mutex);
+    thread_wakeup(main_thread_pid);
+    while (msg_try_receive(&msg) == -1) {}
```
Comment (MichelRottleuthner) [inline tests/xtimer_mutex_lock_timeout/main.c]: first: why not just blocking call of `msg_receive` instead of this loop? second: why use msg send/recv here at all? shouldn't the `thread_sleep() + thread_wakeup()` be enough to lock the mutex in the low prio thread before the main thread continues?

File: tests/xtimer_mutex_lock_timeout/main.c [hunk 3]
```diff
@@ -42,9 +49,38 @@ static const shell_command_t shell_commands[] = {
...
+
+    mutex_unlock(test_mutex);
+    (void)irq_disable();
+    puts("THREAD low prio: exiting low");
+    msg_send_int(&msg, main_thread_pid);
```
Comment (MichelRottleuthner) [inline tests/xtimer_mutex_lock_timeout/main.c]: why `msg_send_int`? This is clearly not called from interrupt context

File: tests/xtimer_mutex_lock_timeout/main.c [hunk 4]
```diff
@@ -114,6 +152,78 @@ static int cmd_test_xtimer_mutex_lock_timeout_long_locked(int argc,
...
+    /* to make the test easier to read */
+    printf("\n");
+
+    return 0;
+}
+
+/**
+ * @brief   shell command to test xtimer_mutex_lock_timeout
+ *
+ * This function will create a new thread with lower prio.
```
Comment (MichelRottleuthner) [inline tests/xtimer_mutex_lock_timeout/main.c]: dot at the end of the line can be removed

File: tests/xtimer_mutex_lock_timeout/main.c (hunk 3 above)
Comment (JulianHolzwarth) [reply tests/xtimer_mutex_lock_timeout/main.c]: to join threads. Only way to wait for a thread to finish that has a lower prio as far as I know(is done this way in posix reaper implementation)

Comment (JulianHolzwarth) [reply tests/xtimer_mutex_lock_timeout/main.c]: `msg_send_int` does not yield but still unblocks the other thread.

File: tests/xtimer_mutex_lock_timeout/main.c (hunk 1 above)
Comment (JulianHolzwarth) [reply tests/xtimer_mutex_lock_timeout/main.c]: Ok.

File: tests/xtimer_mutex_lock_timeout/main.c (hunk 4 above)
Comment (JulianHolzwarth) [reply tests/xtimer_mutex_lock_timeout/main.c]: Ok

File: tests/xtimer_mutex_lock_timeout/main.c (hunk 2 above)
Comment (JulianHolzwarth) [reply tests/xtimer_mutex_lock_timeout/main.c]: after `thread wakeup ` the main thread runs anyways. I don't need that line it is only to better understand what is happening (the thread waits but is still runnable)
The thread should only check receive once.
I can remove it if it is more confusing than helpful.

**Comment by JulianHolzwarth** [comment]
There are more tests I want that are currently failing. tracking #11660
Checking timing should be more done for `xtimer_set64` than this one.
Checking if timeout is triggering is tested in `cmd_test_xtimer_mutex_lock_timeout_long_locked`. (Mutex does not check the owner so locking it in another thread and calling `xtimer_mutex_lock_timeout` should not make any difference)

File: tests/xtimer_mutex_lock_timeout/main.c (hunk 3 above)
Comment (JulianHolzwarth) [reply tests/xtimer_mutex_lock_timeout/main.c]: I will split it in a function with documentation

Comment (JulianHolzwarth) [reply tests/xtimer_mutex_lock_timeout/main.c]: I added a function for it.

**Comment by JulianHolzwarth** [comment]
I want to rebase this pr before it gets merged.

**Comment by JulianHolzwarth** [comment]
@MichelRottleuthner What do you think after the changes and comments I made?

File: tests/xtimer_mutex_lock_timeout/main.c (hunk 2 above)
Comment (JulianHolzwarth) [reply tests/xtimer_mutex_lock_timeout/main.c]: removed it

**Comment by JulianHolzwarth** [comment]
@MichelRottleuthner Could you look at this? What do you think after the changes and comments I made?

File: tests/xtimer_mutex_lock_timeout/main.c [hunk 5]
```diff
@@ -114,6 +172,77 @@ static int cmd_test_xtimer_mutex_lock_timeout_long_locked(int argc,
...
+}
+
+/**
+ * @brief   shell command to test xtimer_mutex_lock_timeout
+ *
+ * This function will create a new thread with lower prio
+ * than the main thread (this function should be called from
+ * the main thread). The new thread will get a mutex and will
+ * lock it. This function (main thread) messages the other
+ * thread and calls xtimer_mutex_lock_timeout.
```
Comment (MichelRottleuthner) [inline tests/xtimer_mutex_lock_timeout/main.c]: This part of the documentation is not valid anymore. i.e. the messaging before calling xtimer_mutex_lock_timeout was removed.

File: tests/xtimer_mutex_lock_timeout/main.c [hunk 6]
```diff
@@ -114,6 +172,77 @@ static int cmd_test_xtimer_mutex_lock_timeout_long_locked(int argc,
...
+ * The other thread will then unlock the mutex. The main
+ * thread gets the mutex and wakes up. The timer will not
+ * trigger because the main threads gets the mutex.
+ *
+ * @param[in] argc  Number of arguments
+ * @param[in] argv  Array of arguments
+ *
+ * @return 0 on success
```
Comment (MichelRottleuthner) [inline tests/xtimer_mutex_lock_timeout/main.c]: I think in this case it can be argued what counts as sucess and what not but the function *always* return 0 even in "error mutex not locked" condition.

File: tests/xtimer_mutex_lock_timeout/main.c [hunk 7]
```diff
@@ -42,9 +49,58 @@ static const shell_command_t shell_commands[] = {
       cmd_test_xtimer_mutex_lock_timeout_long_unlocked, },
     { "mutex_timeout_long_locked", "locked mutex with long timeout",
       cmd_test_xtimer_mutex_lock_timeout_long_locked, },
+    { "mutex_timeout_long_locked_low",
+      "locked mutex from lower prio thread function with long timeout",
```
Comment (MichelRottleuthner) [inline tests/xtimer_mutex_lock_timeout/main.c]: actually the description part of the command is confusing as its not really explaining what happens here. Somewhere in the test it should be explained in clear to understand language what the test checks, what kind of guarantee this implies and by that what it means if the test fails. Maybe for the short description in the command something like "checks if xtimer_mutex_lock_timeout can acquire a mutex that was previously locked by a lower prio thread" would be better to understand for a random dude looking at this test for the first time. Also the terms low and long are not very precise here. I guess you want to express something like *longer than needed* i.e. the timeout is big enough that we are absolutely sure the timeout should never actually be triggered if everything is working fine. And *too short* as in "the timeout is so small that we know it should always trigger if everything works.

**Comment by JulianHolzwarth** [comment]
Working on it Friday.

File: tests/xtimer_mutex_lock_timeout/main.c (hunk 7 above)
Comment (JulianHolzwarth) [reply tests/xtimer_mutex_lock_timeout/main.c]: The description should be something short because a long one is difficult to read. If a "random dude" wants to know what the test does he should read the documentation. I would almost say the command and description is currently too long. I agree that the description is currently not perfect. What do you think about` locked mutex with long timeout tested with 2 threads`
The part: `locked mutex with long timeout` is to be consistent with the other commands.

**LONG** means the timer does not spin(timeout longer than `XTIMER_BACKOFF`).
NOT that is does not trigger. 
It should trigger in `cmd_test_xtimer_mutex_lock_timeout_long_locked`.
It is only called **LONG** because there will be a test with a short one (meaning the timer will spin). This test will be added as soon as the bugs are removed because it fails at the moment.

**tracking pr to show what my goal is (how I will remove the bugs)#11660**

**LOW** and **LONG** are not meant to be precise. It is only short for lower than main thread and longer than `XTIMER_BACKOFF`. How much longer/lower is irrelevant for the test.

To sum it up for me a help description should be short. I am not against changing the  description. And long and low are chosen on purpose.

Comment (MichelRottleuthner) [reply tests/xtimer_mutex_lock_timeout/main.c]: Ok thanks for the clarification. After understanding your intentions for the naming (long) I still think its not precise. I.e. as a reader you don't get the relevant information - that is "no-spin", how about using that instead?. 
>To sum it up for me a help description should be short.

I agree. but it should also be as clear as possible because otherwise its no help ;) How about `lock low-prio-locked-mutex from high-prio-thread (no-spin timeout)`

Comment (MichelRottleuthner) [reply tests/xtimer_mutex_lock_timeout/main.c]: (the dashes are meant to make it more clear which word modifies what)

Comment (JulianHolzwarth) [reply tests/xtimer_mutex_lock_timeout/main.c]: `mutex_timeout_long_locked_low lock low-prio-locked-mutex from high-prio-thread (no-spin timeout)` is a little difficult to read

Comment (JulianHolzwarth) [reply tests/xtimer_mutex_lock_timeout/main.c]: I changed it.

Comment (MichelRottleuthner) [reply tests/xtimer_mutex_lock_timeout/main.c]: True - but if you know where the description starts its not really worse than before. and at least it more unambiguous...

Comment (MichelRottleuthner) [reply tests/xtimer_mutex_lock_timeout/main.c]: Would you also change the "long" part of the other descriptions to "no-spin" to make it consistent again?

Comment (JulianHolzwarth) [reply tests/xtimer_mutex_lock_timeout/main.c]: Yes.

**Comment by MichelRottleuthner** [comment]
please also have a look at the commit message restrictions that Travis is complaining about

**Comment by JulianHolzwarth** [comment]
> please also have a look at the commit message restrictions that Travis is complaining about

**Comment by JulianHolzwarth** [comment]
missclick wanted to comment.

**Comment by JulianHolzwarth** [comment]
> please also have a look at the commit message restrictions that Travis is complaining about
done

**Comment by MichelRottleuthner** [comment]
Hmm seems the test is now too big for some arduino boards and a small nucleo. Do you think we can shrink it down a bit? I think it would preferable to have it working on as many platforms as possible. 
Some things that we could try to reduce code size: 
- comment out the ps and shell_commands module and adding a note that this can be optionally enabled for better manual testing (its not required for the automatic test, right?).
- Remove prints that are not absolutely required, and shrink down those that are. E.g. the string of the test commands itself can be reduced to something like "test1", "test2" etc. since the description already carries all the required information; it also makes reading the shell help easier..
- maybe reuse strings used in multiple places.
- reduce stacksize of thread_low_prio_test

**Comment by JulianHolzwarth** [comment]
I added the same `BOARD_INSUFFICIENT_MEMORY` as the shell test does and reduced memory.

**Comment by JulianHolzwarth** [comment]
Travis fails because of fixups. (rebase after @MichelRottleuthner sees my changes)

**Comment by MichelRottleuthner** [comment]
OK I looked thru the new fixup commits. Feel free to squash and rebase.

**Comment by JulianHolzwarth** [comment]
> OK I looked thru the new fixup commits. Feel free to squash and rebase.

Ok. Done.
